Mathlib/Data/Fin/Basic.lean

```diff
@@ -2045,6 +1328,19 @@ end AddGroup
 
 section SuccAbove
 
+/-- `succAbove p i` embeds `Fin n` into `Fin (n + 1)` with a hole around `p`. -/
+def succAbove (p : Fin (n + 1)) (i : Fin n) : Fin (n + 1) :=
+  if i.1 < p.1 then castSucc i else succ i
+
+/-- `predAbove p i` embeds `i : Fin (n+1)` into `Fin n` by subtracting one if `p < i`. -/
+def predAbove (p : Fin n) (i : Fin (n + 1)) : Fin n :=
+  if h : castSucc p < i then i.pred ((ne_iff_vne i 0).mpr (Nat.not_eq_zero_of_lt h))
+  else i.castLT (Nat.lt_of_le_of_lt (Nat.ge_of_not_lt h) p.2)
+
+/-- `castPred` embeds `i : Fin (n + 2)` into `Fin (n + 1)`
+by lowering just `last (n + 1)` to `last n`. -/
+def castPred (i : Fin (n + 2)) : Fin (n + 1) := predAbove (last n) i
```

Comment:
Should we put these back to how they were in mathlib3, and align them?

Reply:
[#portdatafi](https://github.com/leanprover-community/mathlib4/blob/f4e42872269c9502498e2c2c9e5fae8753936920/Mathlib/Data/Fin/Basic.lean#L2032-L2042)

Reply:
I think these were removed in [this diff](https://github.com/leanprover-community/mathlib4/commit/906f72214116b27baae2f0dd5d7fe6d119c6c373#diff-6ae63812d312aba302a17e36f91f095aee62f00d3cb2f4966f176ea040474db6)

Reply:
Ah, only `succAbove` was changed; I moved the other lemmas back to their original (unbundled) homes.